prereise/gather/demanddata/transportation_electrification/generate_scaling_factors.py

```diff
@@ -0,0 +1,206 @@
+        keep_default_na=False,
+        usecols=[1, 2, 4],
+        skiprows=2,
+    )
+    state2ua = {}
+    for n, s in id2state.items():
+        state2ua[s] = df.query("STATE==@n")["POP"]
+
+    return state2ua
+
+
+def load_census_state(path):
+    """Load census data for state population
+
+    :param str path: path to census file (local or url)
+    :return: (*pandas.Series*) -- Indices are state abbreviations and values are
+        population in state
+    """
+    df = pd.read_excel(path, skiprows=3, index_col=0, skipfooter=7)
+    df.index = df.index.map(lambda x: str(x)[1:])
+
+    return df.loc[state2abv.keys()][2015].rename(index=state2abv)
+
+
+def load_dot_vmt_per_capita(path):
+    """Load Vehicle Miles Traveled (VMY) per capita in urban areas
+
+    :param str path: path to Department of Transportation's Transportation Health Tools
+        (local or url)
+    :return: (*tuple*) -- Series. Indices are state abbreviations and values are
+        population in ua (first element) or state (second element)
+    """
+    df_ua = pd.read_excel(
+        path,
+        sheet_name="Urbanized Area",
+        index_col=0,
+        usecols=[0, 3],
+        names=["UA", "VMT per Capita (daily)"],
+    )
+
+    df_state = pd.read_excel(
+        path,
+        sheet_name="State",
+        index_col=0,
+        usecols=[0, 39],
+        names=["State", "VMT per Capita (annual)"],
+    )
+
+    return (
+        df_ua.squeeze().loc[lambda x: x != "[no data]"],
+        df_state.rename(index=state2abv).squeeze(),
+    )
+
+
+def calculate_vmt_for_ua(census_ua, tht_ua):
+    """Calculate the total annual Vehicle Miles Traveled (VMT) in urban areas
+
+    :param dict census_ua: dictionary as returned by :func:`load_census_ua`
+    :param pandas.Series tht_ua: vmt per capita in urban areas as returned by
+        :func:`load_dot_vmt_per_capita`
+    :return: (*dict*) -- keys are state abbreviation and values are data frames giving
+        annual vmt by urban areas.
+    """
+
+    tht_ua_format = tht_ua.copy()
+    vmt_for_ua = {}
+
+    pattern = [",", "-", ".", " "]
+    for p in pattern:
+        tht_ua_format.index = tht_ua_format.index.str.replace(p, "", regex=True)
+    format2original = dict(zip(tht_ua_format.index, tht_ua.index))
+
+    for s in census_ua:
+        census_ua_format = census_ua[s].copy()
+        for p in pattern:
+            census_ua_format.index = census_ua_format.index.str.replace(
+                p, "", regex=True
+            )
+        common = set(tht_ua_format.index).intersection(set(census_ua_format.index))
+        vmt_for_ua[s] = (
+            pd.DataFrame(
+                {
+                    "Annual VMT": [
+                        365 * tht_ua_format.loc[i] * census_ua_format.loc[i]
+                        for i in common
+                    ]
+                },
+                index=list(common),
+            )
+            .rename(index=format2original)
+            .squeeze()
+        )
```

Comment:
Maybe we can construct a data frame with UA only given the state abbreviations are suffixes of UA names, i.e. `pd.concat(census_ua.values())`, then we should be able to do the multiplication directly with `tht_ua_format`?

Reply:
The problem is that the UA names are not unique. For example you have:
```
>>> vmt_ua = calculate_vmt_for_ua(census_ua, tht_ua)
>>> vmt_ua["OH"]
Cleveland, OH                     1.450011e+10
Toledo, OH-MI                     4.215456e+09
Parkersburg, WV-OH                6.109252e+07
Canton, OH                        2.109291e+09
Middletown, OH                    7.553366e+08
Huntington, WV-KY-OH              2.850411e+08
Springfield, OH                   7.482752e+08
Lorain-Elyria, OH                 2.271439e+09
Wheeling, WV-OH                   3.209512e+08
Youngstown, OH-PA                 3.122444e+09
Dayton, OH                        6.994014e+09
Akron, OH                         5.258336e+09
Columbus, OH                      1.153874e+10
Mansfield, OH                     6.816990e+08
Weirton-Steubenville, WV-OH-PA    3.434335e+08
Lima, OH                          6.162137e+08
Newark, OH                        5.972769e+08
Cincinnati, OH-KY-IN              1.258937e+10
Name: Annual VMT, dtype: float64
>>> vmt_ua["PA"]
Hagerstown, MD-WV-PA              7.916503e+07
Harrisburg, PA                    3.792602e+09
Johnstown, PA                     3.919511e+08
Lebanon, PA                       3.038400e+08
Lancaster, PA                     2.520083e+09
Uniontown-Connellsville, PA       3.988131e+08
York, PA                          1.561355e+09
Cumberland, MD-WV-PA              2.896863e+05
Scranton, PA                      2.841143e+09
Williamsport, PA                  6.703207e+08
Allentown, PA-NJ                  4.718352e+09
Youngstown, OH-PA                 3.541347e+08
Pottstown, PA                     5.967203e+08
Altoona, PA                       5.087245e+08
Reading, PA                       1.739473e+09
Weirton-Steubenville, WV-OH-PA    2.598250e+06
Pittsburgh, PA                    1.207367e+10
Hazleton, PA                      3.577484e+08
State College, PA                 3.945960e+08
Philadelphia, PA-NJ-DE-MD         2.501896e+10
Binghamton, NY-PA                 2.692702e+07
Monessen-California, PA           4.604961e+08
Erie, PA                          1.063796e+09
Name: Annual VMT, dtype: float64
>>> vmt_ua["WV"]
Parkersburg, WV-OH                4.803244e+08
Hagerstown, MD-WV-PA              5.980238e+08
Huntington, WV-KY-OH              9.474756e+08
Wheeling, WV-OH                   5.430394e+08
Morgantown, WV                    4.670725e+08
Charleston, WV                    2.317763e+09
Weirton-Steubenville, WV-OH-PA    2.638600e+08
Cumberland, MD-WV-PA              2.101628e+07
Name: Annual VMT, dtype: float64
```
For example, `Weirton-Steubenville, WV-OH-PA` is one UA (belonging to 3 diiferent states with 3 different values)

Reply:
I thought we will have `Weirton-Steubenville, WV`, `Weirton-Steubenville, OH` and `Weirton-Steubenville, PA` respectively and each (UA, State) pair is unique.

Reply:
The other challenge here is the shapefiles for the spatial translation.  From what Dan Muldrew tested the other day, I believe the UAs in those files are still the whole `Weirton-Steubenville, WV-OH-PA` name, so we did not want to lose the full string, at least for now

Reply:
@BainanXia , that said, you're correct that a cleaner way to do this would be `Weirton-Steubenville, WV`, `Weirton-Steubenville, OH` and `Weirton-Steubenville, PA`
If we can confirm that we won't break other processes, that would certainly be a preference.
We'll indeed have the UA and the specific state in its own column in the csv outputs later, fyi

Reply:
I was thinking we break such multi-state UAs into UA in each state for both population and VMT per capita (even though the latter is the same among each state). Then we will get unique pairs of UA+state so that we could get rid of the multi-layer structure, {state: UA within the state, ...}. But yeah, it doesn't matter as long as it works.